rllib/utils/actor_manager.py

```diff
@@ -0,0 +1,491 @@
+        return self.result_or_error.get()
+
+
+class RemoteCallResults:
+    """Represents a list of results from calls to a set of actors.
+
+    CallResults provides convenient APIs to iterate over the results
+    while skipping errors, etc.
+
+    Example:
+    >>> manager = FaultTolerantActorManager(
+    ...     actors, max_remote_requests_in_flight_per_actor=2,
+    ... )
+    >>> results = manager.foreach_actor(lambda w: w.call())
+    >>>
+    >>> # Iterate over all results ignoring errors.
+    >>> for result in results.ignore_errors():
+    >>>     print(result.get())
+    """
+
+    class _Iterator:
+        """An iterator over the results of a remote call."""
+
+        def __init__(self, call_results: List[CallResult]):
+            self._call_results = call_results
+
+        def __iter__(self) -> Iterator[CallResult]:
+            return self
+
+        def __next__(self) -> CallResult:
+            if not self._call_results:
+                raise StopIteration
+            return self._call_results.pop(0)
+
+    def __init__(self):
+        self.result_or_errors: List[CallResult] = []
+
+    def add_result(self, actor_idx: int, result_or_error: ResultOrError):
+        """Add index of a remote actor plus the call result to the list.
+
+        Args:
+            actor_idx: Index of the remote actor.
+            result_or_error: The result or error from the call.
+        """
+        self.result_or_errors.append(CallResult(actor_idx, result_or_error))
+
+    def __iter__(self) -> Iterator[ResultOrError]:
+        """Return an iterator over the results."""
+        # Shallow copy the list.
+        return self._Iterator(copy.copy(self.result_or_errors))
+
+    def ignore_errors(self) -> Iterator[ResultOrError]:
+        """Return an iterator over the results, skipping errors."""
+        return self._Iterator(
+            [r for r in self.result_or_errors if r.result_or_error.ok]
+        )
+
+
+class FaultTolerantActorManager:
+    """A manager that is aware of the healthiness of remote actors.
+
+    Example:
+    >>> import ray
+    >>> from ray.rllib.utils.actor_manager import FaultTolerantActorManager
+    >>>
+    >>> @ray.remote
+    ... class MyActor:
+    ...    def apply(self, fn) -> Any:
+    ...        return fn(self)
+    ...
+    ...    def do_something(self):
+    ...        return True
+    >>>
+    >>> actors = [MyActor.remote() for _ in range(3)]
+    >>> manager = FaultTolerantActorManager(
+    ...     actors, max_remote_requests_in_flight_per_actor=2,
+    ... )
+    >>>
+    >>> # Synchrnous remote calls.
+    >>> results = manager.foreach_actor(lambda actor: actor.do_something())
+    >>> # Print results ignoring returned errors.
+    >>> print([r.get() for r in results.ignore_errors()])
+    >>>
+    >>> # Asynchronous remote calls.
+    >>> manager.foreach_actor_async(lambda actor: actor.do_something())
+    >>> time.sleep(2) # Wait for the tasks to finish.
```

Comment:
What would happen if I didn't wait? would the fetch function return an empty list?

Reply:
yes

Reply:
yes